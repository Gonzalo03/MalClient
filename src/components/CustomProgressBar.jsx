import PropTypes from "prop-types";
import { useEffect, useState } from "react";
import "../css/animate.css";

export const CustomProgressBar = ({ percent, color, colorAcent, text }) => {
  const [progress, setProgress] = useState(50);

  useEffect(
    () =>
      setTimeout(() => {
        setProgress((percent / 10) * 100 > 100 ? 100 : (percent / 10) * 100);
      }, 50),
    [setProgress, percent]
  );

  return (
    <div
      className={`w-full bg-icon rounded-xl text-xs text-center text-white ${
        progress === 0 && "py-0.5"
      }`}
    >
      <div
        className={`py-0.5 rounded-xl ${
          progress === 0 && "hidden"
        } progress bg-${color}${colorAcent ? `-${colorAcent}` : "-500"}`}
        style={{ width: `${progress <= 10 ? 100 : progress}%` }}
      >
        {`${text} : ${percent}`}
      </div>
      {progress === 0 && `${text} : ${percent}`}
    </div>
  );
};

CustomProgressBar.propTypes = {
  percent: PropTypes.number.isRequired,
  color: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
};

CustomProgressBar.defaultProps = {
  colorAcent: null,
  percent: 100,
};
