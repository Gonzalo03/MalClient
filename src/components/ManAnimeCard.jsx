import { useEffect } from "react";
import PropTypes from 'prop-types'

import { useManime } from "../hook/useManime";

import { CircleLoading } from "react-loadingg";

export const ManAnimeCard = ({ manAnime, context, dispatchManime }) => {
    
  const { loading, data, err } = useManime(manAnime.mal_id, context);
  

  useEffect(
      ()=> data && dispatchManime({type : context, payload : data.score})
  , [dispatchManime, data, context])

  return (
    <>
      {!loading ? (
        <div className="flex rounded overflow-hidden shadow-lg">
          <img className="w-40 object-cover" src={data.image_url} alt={data.title} />
          <div className="px-6 py-4">
            <div className="font-bold text-xl text-custom-first mb-2 w-60">
              {data.title}
            </div>
            <p className="text-gray-700 text-base text-custom-second flex flex-col space-y-2">
              <span>
                Status :
                <span
                  className={`${
                    data.status.includes('Finished') ? 'text-blue-500' : data.status.includes('Hiatus') ? 'text-yellow-500' : 'text-green-500'}`}
                >
                  {data.status}
                </span>
              </span>
              <span>Rank : {data.rank}</span>
              <span>Mean Score : {data.score}</span>
            </p>
          </div>
        </div>
      ) : (
        <>
          {err ? (
            <p>A ocurrido un error {err}</p>
          ) : (
            <div>
              <CircleLoading
                color="#536DFE"
                style={{ positionAbsolute: "none", margin: "auto" }}
              />
            </div>
          )}
        </>
      )}
    </>
  );
};


ManAnimeCard.propTypes = {
  manAnime : PropTypes.object.isRequired,
  context : PropTypes.string.isRequired,
  dispatchManime : PropTypes.func.isRequired
}