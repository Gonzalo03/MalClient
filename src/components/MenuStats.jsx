import { CustomProgressBar } from "./CustomProgressBar";

export const MenuStats = ({ stats }) => {
    return (
        <>
        <p className="text-center text-xl text-white"><span className="bg-primary p-2 rounded-xl">{stats.days_watched ? 'Anime Stats' : 'Manga Stats'}</span></p>
        <div className="flex flex-col space-y-3">
            <CustomProgressBar color='gray' percent={stats.days_watched || stats.days_read} text='Spend days'/>
            <CustomProgressBar color='green' percent={stats.watching || stats.reading} text='Current'/>
            <CustomProgressBar color='blue' colorAcent="700" percent={stats.completed} text='Completed'/>
            <CustomProgressBar color='yellow' percent={stats.on_hold} text='On hold'/>
            <CustomProgressBar color='red' percent={stats.dropped} text='Dropped'/>
            <CustomProgressBar color='blue' percent={stats.mean_score} text='Average'/>
        </div>
        </>
    )
}

