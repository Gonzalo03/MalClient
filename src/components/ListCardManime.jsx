import { ManAnimeCard } from "./ManAnimeCard";

export const ListCardManime = ({ manAnimes, context, dispatchManime }) => {
  return (
    <>
      <p className="text-center text-xl text-white">
        <span className="bg-primary p-2 rounded-xl">
          {context[0].toUpperCase() + context.substring(1)} Favorites
        </span>
      </p>
      <ul className="flex flex-wrap justify-center space-y-6 my-8">
        {manAnimes.map(
          (manAnime, index) =>
            index <= 1 && (
              <ManAnimeCard
                manAnime={manAnime}
                key={manAnime.mal_id}
                context={context}
                dispatchManime={dispatchManime}
              />
            )
        )}
      </ul>
    </>
  );
};
