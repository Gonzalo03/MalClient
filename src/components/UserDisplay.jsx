import { useRef, useEffect, useReducer } from "react";


import { useUser } from "../hook/useUser";
import { useManganimeMemo } from "../hook/useManganimeMemo";
import { avgReducer } from "../reducer/avgReducer";



import { CircleLoading } from "react-loadingg";
import { MenuStats } from "./MenuStats";
import { ListCardManime } from "./ListCardManime";

export const UserDisplay = ({ tag, deleteUser }) => {
  const { loading, data, err } = useUser(tag);

  const [avgManime, dispatchManime] = useReducer(avgReducer, {
    anime : [],
    manga : []
  });

  const computeAvgAnime = useManganimeMemo(avgManime.anime) || 0;
  const computeAvgManga = useManganimeMemo(avgManime.manga) || 0;

  const loadingRef = useRef();
  const aboutRef = useRef();

  useEffect(() => {
    loadingRef.current &&
      loadingRef.current.scrollIntoView({ behavior: "smooth" });

    if (aboutRef.current && data) {
      aboutRef.current.innerHTML = data.about || "";

      setTimeout(() => {
        aboutRef && aboutRef.current.scrollIntoView({ behavior: "smooth" });
      }, 100);
    }
  }, [tag, data]);



  return (
    <>
      {!loading && !err ? (
        <>
          <div className="md:grid grid-cols-2 my-5 userDisplay">
            {/* Profile Preview */}
            <div className="max-w-sm flex flex-col items-center justify-center rounded shadow-lg mx-auto">
              <img src={data.image_url} alt={tag} style={{ width: 300 }} />
              <div className="flex flex-col space-y-5 px-6 pt-4">
                <div className="font-bold text-xl mb-2">{tag}</div>
                <p
                  className="text-icon text-base overflow-hidden space-y-4"
                  ref={aboutRef}
                  style={{ width: 300 }}
                ></p>
                {computeAvgAnime &&<p
                    className={`text-${
                      computeAvgAnime > 7
                        ? "green"
                        : computeAvgAnime < 7 && computeAvgAnime > 4
                        ? "yellow"
                        : computeAvgAnime < 4
                        ? "red"
                        : ""
                    }-500`}
                  >
                    Anime score : {computeAvgAnime} ={">"}{" "}
                    {computeAvgAnime > 7
                      ? "GOD"
                      : computeAvgAnime < 7 && computeAvgAnime > 4
                      ? "Neutral"
                      : computeAvgAnime < 4
                      ? "ZZZZZZ"
                      : ""}
                  </p>}
                {computeAvgManga && (
                  <p
                    className={`text-${
                      computeAvgManga > 7
                        ? "green"
                        : computeAvgManga < 7 && computeAvgManga > 4
                        ? "yellow"
                        : computeAvgManga < 4
                        ? "red"
                        : ""
                    }-500`}
                  >
                    Manga score : {computeAvgManga} ={">"}{" "}
                    {computeAvgManga > 7
                      ? "GOD"
                      : computeAvgManga < 7 && computeAvgManga > 4
                      ? "Neutral"
                      : computeAvgManga < 4
                      ? "ZZZZZZ"
                      : ""}
                  </p>
                )}
              </div>
              <button
                className="bg-primary hover:bg-secondary text-white font-bold py-2 px-4 rounded-full my-5"
                onClick={() => deleteUser(tag)}
              >
                Remover
              </button>
            </div>

            {/* Stats Preview */}

            <div className="flex flex-col space-y-7 self-center mt-10">
              <MenuStats stats={data.anime_stats} />
              <MenuStats stats={data.manga_stats} />
            </div>
          </div>

          {/* Fvorites Preview */}

          <div className="mt-10">
            <ListCardManime
              manAnimes={data.favorites.anime}
              context="anime"
              dispatchManime={dispatchManime}
            />
            <ListCardManime
              manAnimes={data.favorites.manga}
              context="manga"
              dispatchManime={dispatchManime}
            />
          </div>
        </>
      ) : (
        <>
          {err ? (
            <p>A ocurrido un error {err}</p>
          ) : (
            <div ref={loadingRef}>
              <CircleLoading
                color="#536DFE"
                style={{ positionAbsolute: "none", margin: "auto" }}
              />
            </div>
          )}
        </>
      )}
    </>
  );
};
