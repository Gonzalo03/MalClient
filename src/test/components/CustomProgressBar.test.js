import { shallow } from "enzyme"
import { CustomProgressBar } from "../../components/CustomProgressBar"


describe('Pruebas en CustomProgessBar', () => {
    
    
    test('Debe hacer match con el snapshot', () => {
        
        const wrapper = shallow( <CustomProgressBar percent={100} color="blue" text="hold" /> )

        expect(wrapper).toMatchSnapshot();
    })

    
    test('Debe mostrar el accent por default', () => {
        
        const wrapper = shallow( <CustomProgressBar percent={100} color="blue" text="hold" /> )

        const progress = wrapper.find('.progress');

        expect(progress.hasClass('bg-blue-500')).toBeTruthy();

    })

    

})
