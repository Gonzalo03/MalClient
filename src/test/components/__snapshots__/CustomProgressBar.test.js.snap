// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en CustomProgessBar Debe hacer match con el snapshot 1`] = `
<div
  className="w-full bg-icon rounded-xl text-xs text-center text-white false"
>
  <div
    className="py-0.5 rounded-xl false progress bg-blue-500"
    style={
      Object {
        "width": "50%",
      }
    }
  >
    hold : 100
  </div>
</div>
`;
